File: aichain/proof.go
// Copyright (C) 2026, Lux Partners Limited. All rights reserved.
// See the file LICENSE for licensing terms.

package aichain

import (
	"fmt"

	"github.com/luxfi/crypto"
	"github.com/luxfi/geth/common"
)

// MaxProofDepth caps the Merkle path length the verifier (and the precompile)
// will walk — a 2^64-leaf tree is never reached in practice, but an unbounded
// pathLen is a calldata-DoS, so it is hard-capped (LP-5301 MaxProofDepth).
const MaxProofDepth = 64

// proofFrameHeader is the fixed prefix of the LP-5301 proof wire frame:
// ReceiptRoot(32) | u64be Index(8) | u16be pathLen(2).
const proofFrameHeader = 32 + 8 + 2

// MerkleProof is an inclusion proof that a receipt_hash leaf is committed under a
// ReceiptRoot: the sibling hashes from leaf to root plus the leaf Index (which
// fixes left/right at each level). It mirrors chains/aivm MerkleProof and the
// LP-5301 proofBytes frame exactly so a proof exported by the A-Chain
// (Engine.ExportReceipt) verifies here and vice-versa.
type MerkleProof struct {
	ReceiptRoot common.Hash   `json:"receiptRoot"` // the committed root this proof is against
	Index       uint64        `json:"index"`       // leaf index in the tree
	Siblings    []common.Hash `json:"siblings"`    // sibling at each level, leaf->root
}

// leafHashReceipt is the receipt-hash leaf hash: keccak256 over the raw
// receipt_hash digest (one extra keccak so a leaf can never be confused with an
// internal node). Identical to chains/aivm leafHash.
func leafHashReceipt(h common.Hash) common.Hash {
	return common.BytesToHash(crypto.Keccak256(h.Bytes()))
}

// merkleNode combines two child hashes: keccak256(l || r). Identical to
// chains/aivm merkleNode.
func merkleNode(l, r common.Hash) common.Hash {
	return common.BytesToHash(crypto.Keccak256(l.Bytes(), r.Bytes()))
}

// VerifyReceiptInclusion checks that receiptHash is included under root at the
// proof's Index, re-applying the leaf hash and folding with the siblings,
// choosing left/right by the index bit at each level — exactly inverting the
// chains/aivm merkleProof/merkleRoot construction. Returns true iff the
// recomputed root equals root.
func VerifyReceiptInclusion(receiptHash common.Hash, proof MerkleProof, root common.Hash) bool {
	cur := leafHashReceipt(receiptHash)
	idx := proof.Index
	for _, sib := range proof.Siblings {
		if idx%2 == 0 {
			cur = merkleNode(cur, sib)
		} else {
			cur = merkleNode(sib, cur)
		}
		idx /= 2
	}
	return cur == root
}

// Verify is the receipt-aware check: it confirms the proof is against the root
// the proof itself carries (proof.ReceiptRoot) and that the receipt's Hash() is
// included under it. The CALLER must still confirm proof.ReceiptRoot is a root
// the C-Chain holds committed (the precompile does this against its receipt-root
// checkpoint; off-chain you compare against a known-good root).
func (p MerkleProof) Verify(r InferenceReceipt) bool {
	return VerifyReceiptInclusion(r.Hash(), p, p.ReceiptRoot)
}

// EncodeProof serializes a proof into the LP-5301 proofBytes wire frame:
//
//	ReceiptRoot(32) | u64be Index(8) | u16be pathLen(2) | pathLen*32
//
// This is the exact bytes the aivmbridge verifyInferenceReceipt op decodes.
func (p MerkleProof) EncodeProof() ([]byte, error) {
	if len(p.Siblings) > MaxProofDepth {
		return nil, fmt.Errorf("aichain: proof depth %d exceeds max %d", len(p.Siblings), MaxProofDepth)
	}
	out := make([]byte, 0, proofFrameHeader+len(p.Siblings)*32)
	out = append(out, p.ReceiptRoot.Bytes()...)
	out = append(out, u64be(p.Index)...)
	out = append(out, u16be(uint16(len(p.Siblings)))...)
	for _, s := range p.Siblings {
		out = append(out, s.Bytes()...)
	}
	return out, nil
}

// DecodeProof parses an LP-5301 proofBytes frame back into a MerkleProof. It is
// the inverse of EncodeProof and is exact-length: it rejects a short frame, a
// frame whose declared pathLen exceeds MaxProofDepth, and a frame with trailing
// junk — the same calldata-hardening the precompile applies.
func DecodeProof(b []byte) (MerkleProof, error) {
	var p MerkleProof
	if len(b) < proofFrameHeader {
		return p, fmt.Errorf("aichain: proof frame %d bytes, need >= %d", len(b), proofFrameHeader)
	}
	p.ReceiptRoot = common.BytesToHash(b[0:32])
	var idx uint64
	for i := 0; i < 8; i++ {
		idx = idx<<8 | uint64(b[32+i])
	}
	p.Index = idx
	pathLen := int(uint16(b[40])<<8 | uint16(b[41]))
	if pathLen > MaxProofDepth {
		return p, fmt.Errorf("aichain: proof depth %d exceeds max %d", pathLen, MaxProofDepth)
	}
	want := proofFrameHeader + pathLen*32
	if len(b) != want {
		return p, fmt.Errorf("aichain: proof frame %d bytes, want %d for pathLen %d", len(b), want, pathLen)
	}
	p.Siblings = make([]common.Hash, pathLen)
	for i := 0; i < pathLen; i++ {
		off := proofFrameHeader + i*32
		p.Siblings[i] = common.BytesToHash(b[off : off+32])
	}
	return p, nil
}

File: typescript/src/proof.ts
// Copyright (C) 2026, Lux Industries Inc. All rights reserved.
// See the file LICENSE for licensing terms.

import { keccak256, type Hex } from "viem";
import {
  MAX_PROOF_DEPTH,
  bytes32,
  bytesToHex,
  concatBytes,
  u16be,
  u64be,
} from "./bytesutil.js";
import type { InferenceReceipt } from "./receipt.js";
import { receiptHash } from "./receipt.js";

const PROOF_FRAME_HEADER = 32 + 8 + 2; // ReceiptRoot(32) | u64be Index(8) | u16be pathLen(2)

/**
 * MerkleProof is an inclusion proof that a receipt_hash leaf is committed under a
 * receiptRoot. Mirrors chains/aivm MerkleProof and the LP-5301 proofBytes frame.
 */
export interface MerkleProof {
  receiptRoot: Hex;
  index: bigint | number;
  siblings: Hex[];
}

/** leafHash = keccak256(receipt_hash) — identical to chains/aivm leafHash. */
export function leafHash(h: Hex): Hex {
  return keccak256(bytesToHex(bytes32(h)));
}

/** merkleNode = keccak256(l || r) — identical to chains/aivm merkleNode. */
export function merkleNode(l: Hex, r: Hex): Hex {
  return keccak256(bytesToHex(concatBytes(bytes32(l), bytes32(r))));
}

/**
 * verifyReceiptInclusion checks that receiptHash is included under root at the
 * proof's index, exactly inverting the chains/aivm merkle construction.
 */
export function verifyReceiptInclusion(receiptHashHex: Hex, proof: MerkleProof, root: Hex): boolean {
  let cur = leafHash(receiptHashHex);
  let idx = BigInt(proof.index);
  for (const sib of proof.siblings) {
    cur = idx % 2n === 0n ? merkleNode(cur, sib) : merkleNode(sib, cur);
    idx /= 2n;
  }
  return eqHex(cur, root);
}

/** verifyReceipt confirms a receipt's hash is included under the proof's root. */
export function verifyReceipt(proof: MerkleProof, r: InferenceReceipt): boolean {
  return verifyReceiptInclusion(receiptHash(r), proof, proof.receiptRoot);
}

/**
 * encodeProof serializes a proof into the LP-5301 proofBytes wire frame:
 *   ReceiptRoot(32) | u64be Index(8) | u16be pathLen(2) | pathLen*32
 */
export function encodeProof(p: MerkleProof): Uint8Array {
  if (p.siblings.length > MAX_PROOF_DEPTH) {
    throw new Error(`aichain: proof depth ${p.siblings.length} exceeds max ${MAX_PROOF_DEPTH}`);
  }
  return concatBytes(
    bytes32(p.receiptRoot),
    u64be(p.index),
    u16be(p.siblings.length),
    ...p.siblings.map((s) => bytes32(s)),
  );
}

/**
 * decodeProof parses an LP-5301 proofBytes frame. Exact-length: rejects short
 * frames, over-depth, and trailing junk — the same hardening the precompile
 * applies.
 */
export function decodeProof(b: Uint8Array): MerkleProof {
  if (b.length < PROOF_FRAME_HEADER) {
    throw new Error(`aichain: proof frame ${b.length} bytes, need >= ${PROOF_FRAME_HEADER}`);
  }
  const receiptRoot = bytesToHex(b.subarray(0, 32));
  let index = 0n;
  for (let i = 0; i < 8; i++) index = (index << 8n) | BigInt(b[32 + i]!);
  const pathLen = (b[40]! << 8) | b[41]!;
  if (pathLen > MAX_PROOF_DEPTH) {
    throw new Error(`aichain: proof depth ${pathLen} exceeds max ${MAX_PROOF_DEPTH}`);
  }
  const want = PROOF_FRAME_HEADER + pathLen * 32;
  if (b.length !== want) {
    throw new Error(`aichain: proof frame ${b.length} bytes, want ${want} for pathLen ${pathLen}`);
  }
  const siblings: Hex[] = [];
  for (let i = 0; i < pathLen; i++) {
    const off = PROOF_FRAME_HEADER + i * 32;
    siblings.push(bytesToHex(b.subarray(off, off + 32)));
  }
  return { receiptRoot, index, siblings };
}

function eqHex(a: Hex, b: Hex): boolean {
  return a.toLowerCase() === b.toLowerCase();
}
